src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,199 @@
+package pfsdb
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
+)
+
+type (
+	SortOrder      int
+	FilterOperator string
+)
+
+const (
+	SortNone SortOrder = iota
+	SortAscend
+	SortDescend
+)
+
+const (
+	// Note that we use `?` as the placeholder for bindvar, but postgres uses $1, $2, etc.
+	// We solve this by rebinding the query before executing it using sqlx.Rebind.
+	Equal              FilterOperator = "%s = ?"
+	GreaterThan        FilterOperator = "%s > ?"
+	GreaterThanOrEqual FilterOperator = "%s >= ?"
+	LessThan           FilterOperator = "%s < ?"
+	LessThanOrEqual    FilterOperator = "%s <= ?"
+	NotEqual           FilterOperator = "%s <> ?"
+	ValueIn            FilterOperator = "%s in (?)"
+)
+
+type (
+	ModelType interface{ Repo | Commit | Branch }
+	FieldType interface{ CommitField | BranchField }
+	PairType  interface{ CommitPair | BranchPair }
+)
+
+type Filter[T FieldType] struct {
+	Field  T
+	Value  any
+	values []any
+	Op     FilterOperator
+}
+
+func (f *Filter[T]) QueryString() (formatted string, err error) {
+	formatted = fmt.Sprintf(string(f.Op), f.Field)
+	if f.Op == ValueIn {
+		formatted, f.values, err = sqlx.In(formatted, f.Value) // Thank god for this!
+		if err != nil {
+			return "", errors.Wrapf(err, "expanding %s", formatted)
+		}
+	}
+	return
+}
+
+type AndFilters[T FieldType] []*Filter[T]
+
+func (f AndFilters[T]) QueryString() (string, error) {
+	var parts []string
+	for _, filter := range f {
+		part, err := filter.QueryString()
+		if err != nil {
+			return "", err
+		}
+		parts = append(parts, part)
+	}
+	return strings.Join(parts, " AND "), nil
```

Comment:
I think you have to wrap each part in parentheses.  On the go side, binary operations are implicitly grouped like `(and (= foo ?) (= bar ?)` but the generated SQL `foo = ? and bar = ?` only works because `AND` happens to have lower operator precedence than `=`.  If you end up adding OR, then something like `(and (or (= foo ?) (= bar ?)) (or (= baz ?) (= quux ?)))` compiles to `foo = ? OR bar =? AND baz = ? OR quux = ?`, which is interpreted by Postgres as `(or (= foo ?) (and (= bar ?) (= baz ?)) (= quux ?)`, which is not the original expression.

Reply:
One of the reasons I am not a fan of having query builders like this is because you have to pretty quickly end up writing something that support arbitrary cases. I understand that it may be 'bad design' to repeat queries or clauses, but for me having repetition is worth trading off for a query builder.

I understand this is an area where folks may have strong opinions though, but I figured I'd throw in my two cents.